### ISSUE TITLE
Move JUnit4 tests to JUnit5

### ISSUE THREAD

**Issue by nastra** [apache/iceberg#7160]
### Feature Request / Improvement

In our contributing guidelines we point people to writing new unit tests using JUni5. 

We should consider/evaluate what the best approach would be to eventually move existing tests from JUnit4 tests to JUnit5. 

### Query engine

None

**Comment by mitrofmep** [comment]
Hello! I'd like to help with this issue. Just checked [#209](https://github.com/apache/iceberg-docs/pull/209). Am I right, that it needs to check all the test classes, where only "import org.junit.*" used, without import jupiter api? For example, like in `core/src/test/java/org/apache/iceberg/avro/AvroDataTest.java`. And if possible, to refactor them using junit5?

**Comment by nastra** [comment]
It's worth looking at https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4. There are a few items to watch out for:

For example in JUnit4 the method signature of `assertEquals()` was `assertEquals(String message, Object expected, Object actual)` but in JUnit5 it is `assertEquals(Object expected, Object actual, String message)`. There are also a few more methods affected by this. See https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4-failure-message-arguments for some details.
In those cases I'm not sure what's better (just converting from JUnit4 to JUnit5 assertions or switching assertion checks to AssertJ where it makes sense). Both will cause larger diffs.

For modules that don't use JUnit5 yet, those would require
```
test {
    useJUnitPlatform()
  }
```

Things like `@Rule public TemporaryFolder temp = new TemporaryFolder();` don't work in JUni5 and would have to be converted to `@TempDir Path temp;`.

It's worth doing the conversion module-by-module and starting in a module that has a small amount of tests (like `iceberg-pig` or `iceberg-gcp`).

@jackye1995 any thoughts on the outlined approach?

**Comment by jackye1995** [comment]
Yes we definitely want to convert module by module. 

I do not have strong opinion regarding using Junit `assertEquals` vs AssertJ, it seems like they will incur almost the same amount of changes:

```
Assertions.assertThat("a").isEqualTo("b").describedAs("xxx");
```

vs 

```
Assertions.assertEquals("a", "b", "xxx")
```

AssertJ might have the advantage that it can format the description with `.describedAs("xxx %s %s", x, y)` so it's a bit easier to use. And it might simplify some test cases. I don't quite get why we want to lazily load a description message with the feature that Junit5 provides. So maybe using AssertJ could be preferred.

**Comment by jackye1995** [comment]
I just noticed one thing, given the fact that Iceberg prefers to not import static methods directly, we will end up having 2 `Assertions` classes if we use both Junit5 and AssertJ. Given that we are already using AssertJ, we probably need to stick with that instead of mixing both.

This means we want to migrate to Junit5 for framework and usages of things like tempdir, before, after, etc. but use AssertJ for all assertions. Should we have a thread on devlist just to confirm this as a community guideline?

> there are scripts available to convert assertions

That's super cool, maybe we should try it.

**Comment by jackye1995** [comment]
Thank you Weijie for the detailed explanation, this is super helpful! Regarding Junit5 vs AssertJ `Assertions`, what did Flink community do? Do you think using AssertJ `Assertions` (which seems to be the consensus above) is preferred over Junit5?

**Comment by reswqa** [comment]
Thanks @jackye1995 for the quick reply!

> Regarding Junit5 vs AssertJ Assertions, what did Flink community do?
Earlier, many different types of assertions were mixed in Flink's codebase, such as `Junit4` assertion, `Hamcrest`,`some custom assertion`, etc. 

While migrating to `Junit5`, Flink community decided that use `AssertJ` as the only assertion framework.

> Do you think using AssertJ Assertions (which seems to be the consensus above) is preferred over Junit5?
In my personal experience, `AssertJ` is really powerful and the benefit speaks for itself.

- Thanks to its chained assertion style, we can avoid writing a lot of redundant code.
- It has truly helpful error messages.
- It provides a dedicated assertion API for many types, which will greatly reduce the burden of writing tests and increase the readability of the code.
- Provides assertions for Guava types like Multimap, Table, Optional or ByteSource.

**Comment by attilakreiner** [comment]
Hey Team, I am new here, I was looking for a beginner task and I've seen you are in the process of migrating from Junit 4 to 5.

I just gave it a try and migrated over the tests in the :iceberg-data module. (The `:iceberg-flink` module also got affected; and a little bit `:iceberg-spark` and `:iceberg-mr` too). I believe I have completely removed Junit 4 from the `:iceberg-data` module.

Pls review this PR: https://github.com/apache/iceberg/pull/10657

I'd highly appreciate any feedback.

**Comment by tomtongue** [comment]
@nastra I believe the remaining migration is only in Spark 3.4. If possible, can I proceed them to the JUnit 5 with AssertJ style? (and, I think this also should help the backport of V3 spec features to Spark 3.4)
Spark 3.3 is still based in JUnit 4, but it's recently removed from the repository via https://github.com/apache/iceberg/pull/12279. So I think Spark 3.4 migration is enough.

**Comment by nastra** [comment]
@tomtongue yes I believe only Spark 3.4 is left, but it wasn't a priority to migrate it to JUnit5 (since it's a big task). However, feel free to work on this as that will definitely make it easier to backport code to Spark 3.4

**Comment by tomtongue** [comment]
@nastra Thank you. Yes, it's still a big task and I understand the current prioritization. I will start it again gradually from adding the base tests like other versions for the further help.

**Comment by nastra** [comment]
Thanks everyone for contributing on this issue. This was a true community effort and we wouldn't be able to achieve this without so many contributions from all of you 🚀
